pytorch_lightning/plugins/training_type/training_type_plugin.py

```diff
@@ -284,3 +284,14 @@ def call_configure_sharded_model_hook(self) -> bool:
     @call_configure_sharded_model_hook.setter
     def call_configure_sharded_model_hook(self, mode: bool) -> None:
         self._call_configure_sharded_model_hook = mode
+
+    @property
+    def sharded_module(self) -> torch.nn.Module:
+        """
+        Override to pass a sharded version of the module. This is useful for configuring optimizers and training
+        with the sharded version of the ``LightningModule``.
+
+        Returns: The Sharded version of the ``LightningModule`` if using a Sharded Plugin,
+            else the model stored in the training type plugin.
+        """
+        return self.model
```

Comment:
Currently the default is `self.model`. In single device, this is just the lightning module. Using DDP, this is the `DistributedDataParallel` wrapper. In Sharded Training, this is the `ShardedDataParallel` wrapper. In Fully Sharded, this will be the `FullyShardedDataParallel` wrapper.

I went for the wrapper class rather than the lightning module for everything except FSDP because it standardizes across DDP/Sharded/Fully Sharded, and enforces the same logic (you can access the module via model.module, but be wary).

Reply:
isn't it misleading if the user code contains a call to `self.sharded_module` and the rest of the code is plugin/device agnostic?